src/pipecat/processors/canonical_metrics_processor.py

```diff
@@ -0,0 +1,202 @@
+import os
+import uuid
+import wave
+from datetime import datetime
+from io import BytesIO
+from typing import Dict, List, Tuple
+
+import aiohttp
+from loguru import logger
+
+try:
+    import aiofiles
+    import aiofiles.os
+except ModuleNotFoundError as e:
+    logger.error(f"Exception: {e}")
+    logger.error(
+        "In order to use Canonical Metrics, you need to `pip install pipecat-ai[canonical]`. " +
+        "Also, set the `CANONICAL_API_KEY` environment variable.")
+    raise Exception(f"Missing module: {e}")
+
+
+from pipecat.frames.frames import CancelFrame, EndFrame, Frame
+from pipecat.processors.audio_buffer_processor import AudioBufferProcessor
+from pipecat.processors.frame_processor import FrameDirection
+
+"""
+This class extends AudioBufferProcessor to handle audio processing and uploading
+for the Canonical Voice API.
+"""
+
+
+class CanonicalMetrics(AudioBufferProcessor):
```

Comment:
I feel maybe this could be probably an `AIService` instead and we should then pass the `AudioBufferProcessor` as an argument.

So:

```
class CanonicalMetricsService(AIService):

    def __init__(self, audio_bufer_processor: AudioBufferProcessor, ...):
        ...
```

So `CanonicalMetricsServices` would live in `src/pipecat/services` with this approach. Another benefit is that you could then use any other `AudioBufferProcessor` implementation.

The only thing is that you then need to put both processors in the pipeline. But it's probably fine. So,

```
Pipeline([transport.input(), audio_buffer_processor, canonical, ....])
```

Another benefit is that other processors in the pipeline could use the `audio_buffer_processor`.

WDYT?

Reply:
I like it. I'll make the change.

Reply:
Done.

Reply:
Awesome! Thank you. Looking great!